src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -39,67 +41,40 @@ private CSharpSyntaxNode ParseTypeOrPattern()
                 {
                     TypeSyntax type = this.ParseType(ParseTypeMode.AfterIsOrCase);
 
-                    tk = this.CurrentToken.ContextualKind;
-                    if (!type.IsMissing)
+                    if (!type.IsMissing && this.IsTrueIdentifier())
                     {
-                        if (this.IsTrueIdentifier())
-                        {
-                            var designation = ParseSimpleDesignation();
-                            node = _syntaxFactory.DeclarationPattern(type, designation);
-                        }
+                        var designation = ParseSimpleDesignation();
+                        return _syntaxFactory.DeclarationPattern(type, designation);
                     }
 
-                    if (node == null)
+                    tk = this.CurrentToken.ContextualKind;
+                    if ((!IsExpectedBinaryOperator(tk) || GetPrecedence(SyntaxFacts.GetBinaryExpression(tk)) <= precedence) &&
+                        // member selection is not formally a binary operator but has higher precedence than relational
```

Comment:
An is pattern expression is a kind of relational expression, however, "relational expression" refers to a precedence level.